### PR TITLE
feat: encode / decode predicate

### PR DESCRIPTION
**Pull request by freesig**
# Adds 
- predicate encoding, decoding
- program encoding, decoding
- programs encoding, decoding
- hash predicate using encoding

### CODE REVIEW

File: crates/types/src/predicate/encode_predicate.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
+//! # Encode and Decode Predicates
+//!
+//! # Encoding
```
Comment (supiket) [inline crates/types/src/predicate/encode_predicate.rs:3]: Here, should there be a guide for program encoding too?

File: crates/types/src/predicate/encode_predicate.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+//! | Field | Size (bytes) | Description |
+//! | --- | --- | --- |
+//! | edge_start | 2 | The index of the first edge in the edge list. |
+//! | program_address | 32 | The address of the program. |
+//! | reads | 1 | The type of state this program has access to. |
+//!
+//! ## Edge
+//! | Field | Size (bytes) | Description |
+//! | --- | --- | --- |
+//! | edge | 2 | The index of the node that this edge points to. |
```
Comment (supiket) [inline crates/types/src/predicate/encode_predicate.rs:22]: I'm guessing the generated docs look better, but for these comments, do you know if there's a way to easily align the columns?

**Review by supiket — APPROVED** [review]
After looking at the tests, I see your point about being able to inspect the data through ASM (and Pint). It's beautiful.

File: crates/types/src/predicate/encode_predicate.rs (hunk 2 above)
Comment (freesig) [reply crates/types/src/predicate/encode_predicate.rs:22]: Yeh this generates tables in docs. I don't know of a way to align automatically

**Comment by freesig** [comment]
closes #200

**Pull request merged**